ocmsrnaseq/pipeline_fastqc.py

```diff
@@ -110,8 +111,116 @@ def build_report(infiles, outfile):
     statement = '''multiqc . -s'''
     P.run(statement)
 
-@follows(build_report)
+multiqc_file = os.path.abspath("multiqc_data/multiqc_fastqc.txt")
+
+@transform(multiqc_file,  # Use absolute path
+           regex(r".*/multiqc_fastqc.txt"),
+           r"multiqc_data/failed_qc_metrics/")
+def find_failed_samples(infile, outfile):
+    '''Identify samples that failed specific QC metrics and generate separate files for each metric.'''
+    import os
+    import pandas as pd
+
+    # Read the MultiQC FastQC file
+    df = pd.read_csv(infile, sep="\t")
+
+    # List of QC metrics from column headers
+    qc_metrics = df.columns[12:23]  # Columns 13-23 (zero-based index)
+
+    # Create directory to store output files
+    output_dir = os.path.dirname(outfile)
+    os.makedirs(output_dir, exist_ok=True)
+
+    # Loop through each QC metric and generate a separate file
+    for metric in qc_metrics:
+        # Filter samples that failed the QC metric
+        failed_samples = df[df[metric].str.contains("fail", case=False, na=False)]["Sample"]
+
+        # Only create a file if there are failed samples
+        if not failed_samples.empty:
+            output_file = os.path.join(output_dir, f"failed_{metric}.txt")
+            failed_samples.to_csv(output_file, index=False, header=False)
+
+
+@follows(find_failed_samples)
+@merge("multiqc_data/failed_qc_metrics/*.txt", "multiqc_data/failed_qc_metrics/failed_combined_samples.txt")
+def combine_failed_samples(infiles, outfile):
+    '''Combine all failed samples across QC metrics and remove duplicates.'''
+    # Ensure the directory for the output file exists
+    output_dir = os.path.dirname(outfile)
+    os.makedirs(output_dir, exist_ok=True)
```

Comment:
making output directories is unneccessary here as has been made in the previous function